README.md
DiffPeaks
=========

Compute the difference between two peak tracks by alignment


NEWS
TODO

error checking: do the input peaks make sense? are they
non-intersecting?

2014.5.27

first implementation from airplane notes.

tests/testthat.R
library(testthat)
test_check("DiffPeaks")


tests/testthat/test-examples.R
context("Examples")

expect_peaks_diff <- function(peaks1, peaks2, diff.chromStart, diff.chromEnd){
  gen.list <-
    list(matched_diff(peaks1, peaks2),
         matched_diff(peaks2, peaks1))
  for(generated in gen.list){
    expect_identical(generated$chromStart, diff.chromStart)
    expect_identical(generated$chromEnd, diff.chromEnd)
  }
}

test_that("two peaks, no overlap, two differences", {
  expect_peaks_diff(Peaks(1, 5),
                    Peaks(7, 10),
                    c(1L, 7L),
                    c(5L, 10L))
})

test_that("two differences and two matches", {
  expect_peaks_diff(Peaks(c(1, 11, 25), c(4, 20, 30)+1),
                    Peaks(c(6, 11, 21), c(10, 14, 30)+1),
                    c(1L, 6L),
                    c(5L, 11L))
})

test_that("two peaks inside 1 peak -- pick the bigger one", {
  expect_peaks_diff(Peaks(1L, 11L),
                    Peaks(c(2L, 7L), c(5L, 9L)),
                    7L,
                    9L)
})

test_that("one peak completely overlapped, other partially", {
  expect_peaks_diff(Peaks(c(1L, 22L), c(21L, 25L)),
                    Peaks(10L, 25L),
                    22L,
                    25L)
})

test_that("several overlaps", {
  expect_peaks_diff(Peaks(c(10, 100, 210, 230), c(20, 200, 220, 300)),
                    Peaks(c(90, 120, 250, 400), c(110, 240, 300, 500)),
                    c(10L, 90L, 210L, 400L),
                    c(20L, 110L, 220L, 500L))
})


src/matched_diff.h
#include <cstdio>
#include <list>
#include <map>
#include <set>

class Match;
typedef std::multimap< double, Match*, std::greater<double> > SortedMatches;
typedef std::pair< double, Match* > JaccardMatch;
typedef std::map< int, SortedMatches::iterator > MatchMap;
typedef std::pair< int, SortedMatches::iterator > IDMatch;
typedef std::set<int> IDSet;

class Peak;
typedef std::list< Peak* > PeakList;

class Match {
public:
    PeakList::iterator one;
    PeakList::iterator two;
    Match(PeakList::iterator o, PeakList::iterator t){
	one = o;
	two = t;
    }
};

class Peak {
    int chromStart;
    int chromEnd;
    int track;
public:
    MatchMap matches;
    int get_chromStart(){
	return chromStart;
    }
    int get_chromEnd(){
	return chromEnd;
    }
    int get_track(){
	return track;
    }
    Peak(int s, int e, int t){
	chromStart = s;
	chromEnd = e;
	track = t;
    }
};

int matched_diff(int*, int*, int*, int, int*, int*);

#define ERROR_UNRECOGNIZED_PEAK_TRACK -1

class PeakMatches {
    SortedMatches possible_matches;
    IDSet matched;
    int peak_list_chromEnd, new_match_id;
    PeakList::iterator last_one_it, last_two_it;
public:
    PeakList peak_list;
    void init(){
	last_one_it = last_two_it = peak_list.end(); // NULL.
	peak_list_chromEnd = 0;
	new_match_id = 0;
	matched.clear();
    }
    int get_chromEnd(){
	return peak_list_chromEnd;
    }
    int add_peak(Peak *peak){
	PeakList::iterator peak_it;
	int before, after;
	before = peak_list.size();
	peak_list.push_front(peak);
	after = peak_list.size();
	peak_it = peak_list.begin();
	if(peak_list_chromEnd < peak->get_chromEnd()){
	    peak_list_chromEnd = peak->get_chromEnd();
	}
	int track = peak->get_track();
	/* printf("add_peak %d-%d track=%d chromEnd=%d size=%d\n", */
	/*        peak->get_chromStart(), */
	/*        peak->get_chromEnd(), */
	/*        track, peak_list_chromEnd, */
	/*        peak_list.size()); */
	switch(track) {
	case 1:
	    add_match(last_two_it, peak_it);
	    last_one_it = peak_it;
	    break;
	case 2:
	    add_match(last_one_it, peak_it);
	    last_two_it = peak_it;
	    break;
	default:
	    return ERROR_UNRECOGNIZED_PEAK_TRACK;
	}
	return 0;
    }
    void add_match(PeakList::iterator last_it, PeakList::iterator peak_it){
	if(last_it == peak_list.end()){
	    return; // there is no previous peak in this group.
	}
	Peak *peak, *last;
	peak = *peak_it;
	last = *last_it;
	int Union, Intersection, leftEnd, rightEnd;
	if(peak->get_chromEnd() < last->get_chromEnd()){
	    rightEnd = last->get_chromEnd();
	    leftEnd = peak->get_chromEnd();
	}else{
	    rightEnd = peak->get_chromEnd();
	    leftEnd = last->get_chromEnd();
	}
	Union = rightEnd - last->get_chromStart();
	Intersection = leftEnd - peak->get_chromStart();
	double jaccard = (double)Intersection/(double)Union;
	/* printf("%d-%d %d-%d union=%d intersection=%d jaccard=%f\n", */
	/*        last->get_chromStart(), */
	/*        last->get_chromEnd(), */
	/*        peak->get_chromStart(), */
	/*        peak->get_chromEnd(), */
	/*        Union, Intersection, jaccard); */
	Match *match;
	match = new Match(last_it, peak_it);
	JaccardMatch jm = JaccardMatch(jaccard, match);
	SortedMatches::iterator sorted_it = possible_matches.insert(jm);
	IDMatch id_match = IDMatch(new_match_id, sorted_it);
	new_match_id++;
	last->matches.insert(id_match);
	peak->matches.insert(id_match);
    }
    void remove_matching(){
	Match *match;
	JaccardMatch jm;
	SortedMatches::iterator match_it;
	while(!possible_matches.empty()){
	    match_it = possible_matches.begin();
	    jm = *match_it;
	    match = jm.second;
	    /* printf("remove_matching Jaccard %f %d-%d %d-%d\n", jm.first, */
	    /* 	   (*match->one)->get_chromStart(), */
	    /* 	   (*match->one)->get_chromEnd(), */
	    /* 	   (*match->two)->get_chromStart(), */
	    /* 	   (*match->two)->get_chromEnd()); */
	    remove_peak(match->one);
	    remove_peak(match->two);
	}
    }
    void remove_peak(PeakList::iterator peak_it){
	MatchMap::iterator matches_it;
	SortedMatches::iterator sorted_it;
	IDMatch id_match;
	JaccardMatch jm;
	int id;
	Peak *peak;
	Match *match;
	peak = *peak_it;
	/* printf("begin remove_peak peak_list=%d peak->matches=%d possible_matches=%d matched=%d\n",  */
	/*        peak_list.size(), peak->matches.size(),  */
	/*        possible_matches.size(), matched.size()); */
	while(!peak->matches.empty()){
	    matches_it = peak->matches.begin();
	    id_match = *matches_it;
	    id = id_match.first;
	    sorted_it = id_match.second;
	    if(matched.find(id) == matched.end()){
		//printf("marking id==%d\n", id);
		matched.insert(id); // mark as already matched.
		jm = *sorted_it;
		match = jm.second;
		delete match;
		possible_matches.erase(sorted_it);
	    }
	    peak->matches.erase(matches_it);
	}
	peak_list.erase(peak_it);
	delete peak;
	/* printf("end   remove_peak peak_list=%d peak->matches=%d possible_matches=%d matched=%d\n",  */
	/*        peak_list.size(), peak->matches.size(),  */
	/*        possible_matches.size(), matched.size()); */
    }
    bool has_peaks(){
	return !peak_list.empty();
    }
    Peak* pop_peak(){
	PeakList::iterator peak_it;
	Peak *peak;
	peak_it = peak_list.begin();
	peak = *peak_it;
	peak_list.erase(peak_it);
	return peak;
    }
};


src/matched_diff.cpp
 #include <cstdio>
// #include <limits>
//#include <iostream>
#include "matched_diff.h"

int
matched_diff
(int *peak_chromStart,
 int *peak_chromEnd,
 int *peak_track,
 int peak_count,
 int *diff_chromStart,
 int *diff_chromEnd){
  int peak_i = 0, diff_i = 0, status;
  PeakMatches peak_matches;
  Peak *peak;
  while(peak_i < peak_count) {
    peak_matches.init();
    // Add peaks and matches until there are no more overlapping
    // peaks.
    while(peak_matches.get_chromEnd() == 0 || 
	  (peak_i < peak_count && 
	   peak_chromStart[peak_i] < peak_matches.get_chromEnd())){
      peak = new Peak(peak_chromStart[peak_i], 
		      peak_chromEnd[peak_i],
		      peak_track[peak_i]);
      status = peak_matches.add_peak(peak);
      peak_i++;
      if(status < 0){
	return status;
      }
    }
    peak_matches.remove_matching();
    while(peak_matches.has_peaks()){
      peak = peak_matches.pop_peak();
      diff_chromStart[diff_i] = peak->get_chromStart();
      diff_chromEnd[diff_i] = peak->get_chromEnd();
      delete peak;
      diff_i++;
    }
    //printf("overlap done diff_i=%d peak_i=%d\n", diff_i, peak_i);
  }
  
  // 249250621 for hg19:chr1.
  // 248956422 for hg38:chr1.
  // 2147483647 returned by std::numeric_limits<int>::max().
  // 2000000000 == 2e9L in R (integer implemented as long).
  // So this code should work fine.

  // int imin = std::numeric_limits<int>::min(); // minimum value
  // int imax = std::numeric_limits<int>::max();
  // printf("%d %d\n", imin, imax);
   
    return diff_i;
}


R/Peaks.R
### Compute the difference between peak lists.
matched_diff <- structure(function(x, y){
  for(df in list(x, y)){
    stopifnot(is.data.frame(df))
    for(col.name in c("chromStart", "chromEnd")){
      stopifnot(col.name %in% names(df))
      stopifnot(is.integer(df[, col.name]))
    }
  }
  if(nrow(x)==0)return(y)
  if(nrow(y)==0)return(x)
  df <-
    rbind(data.frame(x, track=1L),
          data.frame(y, track=2L))
  df <- df[order(df$chromStart), ]
  diff.df <- matched_diff_interface(df)
  diff.df[order(diff.df$chromStart),]
}, ex=function(){
  some.peaks <- Peaks(c(10, 100, 210, 230), c(20, 200, 220, 300))
  more.peaks <- Peaks(c(90, 120, 250, 400), c(110, 240, 300, 500))
  peak.diff <- matched_diff(some.peaks, more.peaks)
  segs <-
    rbind(data.frame(some.peaks, what="one sample"),
          data.frame(more.peaks, what="another sample"),
          data.frame(peak.diff, what="difference"))
  library(ggplot2)
  ggplot()+
    geom_segment(aes(chromStart, what, xend=chromEnd, yend=what), data=segs)
})

### Make a data.frame representing a list of peaks on 1 chromosome.
Peaks <- function(chromStart, chromEnd){
  stopifnot(chromStart < chromEnd)
  data.frame(chromStart=as.integer(chromStart),
             chromEnd=as.integer(chromEnd))
}
